### ISSUE TITLE
GPU utalization

### ISSUE THREAD

**Issue by jdohn15** [umyelab/LabGym#8]
Hello,

This is similar to issue #7, but I do not feel my quesiton/issue was answered. Analyzing Behavior is not utilizing my GPU it seems.

Thanks!

**Comment by yujiahu415** [comment]
Hi,

Can you send me a screenshot of your CMD prompt that indicates the GPU is not used, here or to henryhu@umich.edu? Thanks!

The Analyzing Behavior should use GPU if: 1. your Windows has a Nvidia GPU, 2. the CUDA toolkit is properly installed, 3. all the files needed to run CUDA (e.g., cudart64_110.dll) are in the Windows\System32 folder.

**Comment by jdohn15** [comment]
I have re-installed CUDA toolkit to make sure. I also tried with a new video, and the ram usage increases quickly to 100% and locks up my computer too. The 4% GPU usage is background stuff.

Thank you for the fast reply!
![image](https://user-images.githubusercontent.com/78576051/184459982-6f1ef7bf-d8ce-41e8-946d-941766a55cf3.png)

**Comment by yujiahu415** [comment]
Hi,

I guess the program was stuck at the stage of extracting background, although I'm not sure--if you can screenshot the CMD prompt, that will help me to understand what / where / when was the issue. Extracting background does not use GPU but uses RAM. Only categorizing behaviors in which the neural network is utilized uses GPU. If the program was indeed stuck at the stage of extracting background, you may try to use a shorter time window for background extraction, or use option 'proportional resize the video frame' to reduce the frame size, which would significantly increase the processing speed and reduce the RAM usage. Typically, a 30-second time window during which the animal is moving around (if it's immobile, it will be considered as static background) is already enough for background extraction. And a 500 x 500 frame size is already of high resolution for behaviors of a single animal unless the animal has a lot fine features that are critical for the behavior categorization.

I'm not sure if I explained clearly enough but just let me know if it's confusing. You may also send me a sample video of yours (via Google Drive or Dropbox) so that I can take a look and provide specific suggestions.

**Comment by jdohn15** [comment]
I ended up getting my GPU to work, thank you for your help! I was using anaconda to install LabGym, but re installed it with python via the Microsoft store. 

As a side note, during "Acquiring information in each frame" how do I speed this up since it seems to be cpu driven?

**Comment by yujiahu415** [comment]
The 'acquiring information in each frame' is a sequential process, which would not be benefited by using GPU. So it's basically cpu-driven and uses RAM. It goes through the video (the duration for analysis), gets the information in each frame, and stores them in RAM, so that they can be made into tensors for later GPU use in 'categorizing behaviors'.

To speed up this step, you may try to reduce the frame size by using the option 'proportionally resize the video frames' (e.g. to 1/2 of its original size, or even smaller). This would significantly increase the processing speed without hurting the categorization accuracy in most of cases. Any segmented animal blob will be downsized to fit the input shape of the network used for behavioral categorization (the Categorizer). For example, if you use the pretrained Categorizer 'Mouse_Homecage_sideview_30fps', the segmented animal blob will be downsized to 64x64 during behavioral categorization. Anything larger than that will not improve the categorization accuracy but just decrease the processing speed. You may have an estimation how big the animal blob is to the video frame size to determine the size for resizing.

Hope this can be helpful.
